### PR TITLE
8303965: java.net.http.HttpClient should reset the stream if response headers contain malformed header fields

### PR DESCRIPTION
**Pull request by dfuch**
According to RFC 9113:

A malformed request or response is one that is an otherwise valid sequence of HTTP/2 frames but is invalid due to the presence of extraneous frames, prohibited fields or pseudo-header fields, the absence of mandatory pseudo-header fields, the inclusion of uppercase field names, or invalid field names and/or values (in certain circumstances; see [Section 8.2](https://www.rfc-editor.org/rfc/rfc9113#HttpHeaders)).
[...]
Malformed requests or responses that are detected MUST be treated as a [stream error](https://www.rfc-editor.org/rfc/rfc9113#StreamErrorHandler) ([Section 5.4.2](https://www.rfc-editor.org/rfc/rfc9113#StreamErrorHandler)) of type [PROTOCOL_ERROR](https://www.rfc-editor.org/rfc/rfc9113#PROTOCOL_ERROR).

The current behavior is to close the connection with protocol error. This change makes it reset the stream instead.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303965](https://bugs.openjdk.org/browse/JDK-8303965): java.net.http.HttpClient should reset the stream if response headers contain malformed header fields

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12976/head:pull/12976` \
`$ git checkout pull/12976`

Update a local copy of the PR: \
`$ git checkout pull/12976` \
`$ git pull https://git.openjdk.org/jdk pull/12976/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12976`

View PR using the GUI difftool: \
`$ git pr show -t 12976`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12976.diff">https://git.openjdk.org/jdk/pull/12976.diff</a>

</details>

### CODE REVIEW

**Comment by dfuch** [comment]
HttpClient tests are stable and tier2 passed

File: src/java.net.http/share/classes/jdk/internal/net/http/Stream.java [hunk 1]
```diff
@@ -1589,15 +1590,28 @@ void reset() {
...
+                String n = name.toString();
+                String v = value.toString();
+                super.onDecoded(n, v);
+                responseHeadersBuilder.addHeader(n, v);
+                if (Log.headers() && Log.trace()) {
+                    Log.logTrace("RECEIVED HEADER (streamid={0}): {1}: {2}",
+                            streamid, n, v);
+                }
+            } catch (UncheckedIOException uio) {
+                // reset stream: From RFC 7540, section-8.1.2.6
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:1603]: Hello Daniel, should we instead refer to RFC-9113 (section 8.1.1) which obsoletes RFC-7540?

File: src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java [hunk 2]
```diff
@@ -1204,6 +1208,21 @@ private void sendConnectionPreface() throws IOException {
...
+    /**
+     * Called do get the initial stream after a connection upgrade.
+     * If the stream was cancelled, it might no longer be in the
+     * stream map. Therefore - we use the initialStream field
+     * instead, and reset it to null after returning it.
+     * @param <T> the response type
+     * @return the initial stream created during the upgrade.
+     */
+    @SuppressWarnings("unchecked")
+    <T> Stream<T> getInitialStream() {
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:1220]: Given that this method returns and also updates the initial stream member field, the naming of this method is a bit odd. But I can't think of a better name, plus this is internal to the `jdk.internal.net.http` package and also has a comment which explains what it does, so I think this name is fine.

File: src/java.net.http/share/classes/jdk/internal/net/http/common/ValidatingHeadersConsumer.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ * questions.
+ */
+package jdk.internal.net.http.common;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Set;
+
+/*
+ * Checks RFC 7540 rules (relaxed) compliance regarding pseudo-headers.
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/common/ValidatingHeadersConsumer.java:32]: Same here, should we use new RFC number?

**Review by jaikiran — APPROVED** [review]
The change looks good to me. I just have a question about the RFC numbers in the comment, which I have asked inline.

File: src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java (hunk 2 above)
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:1220]: Maybe I could change the name to "retrieveInitialStream()"... Would that be better?

File: src/java.net.http/share/classes/jdk/internal/net/http/Stream.java [hunk 4]
```diff
@@ -1589,15 +1590,28 @@ void reset() {
...
+                // Malformed requests or responses that are detected MUST be
+                // treated as a stream error (Section 5.4.2) of type
+                // PROTOCOL_ERROR.
+                onProtocolError(uio.getCause());
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:1607]:
```suggestion
                // reset stream: From RFC 9113, section 8.1
                // Malformed requests or responses that are detected MUST be
                // treated as a stream error (Section 5.4.2) of type
                // PROTOCOL_ERROR.
                onProtocolError(uio.getCause());
```

File: src/java.net.http/share/classes/jdk/internal/net/http/Stream.java (hunk 1 above)
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:1603]: Right. Sorry for the oversight.

File: src/java.net.http/share/classes/jdk/internal/net/http/common/ValidatingHeadersConsumer.java (hunk 3 above)
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/common/ValidatingHeadersConsumer.java:32]:
```suggestion
 * Checks RFC 9113 rules (relaxed) compliance regarding pseudo-headers.
```

Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/common/ValidatingHeadersConsumer.java:32]: Done

File: src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java (hunk 2 above)
Comment (jaikiran) [reply src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:1220]: Hello Daniel, I don't have strong preference, `retrieveInitialStream` sounds OK. It's also OK if you would want to continue with the current name.

**Review by jaikiran — APPROVED** [review]

**Pull request closed without merging**